lib/auth/oidc/provider.go

```diff
@@ -25,6 +25,13 @@ func providerConfig(authMethod *structs.ACLAuthMethod) (*oidc.Config, error) {
 		algs = []oidc.Alg{oidc.RS256}
 	}
 
+	// if client assertion is enabled, do not send a client secret normally;
+	// if it is set to anything, it will be used as an HMAC to sign the client
+	// assertion JWT, instead.
+	if authMethod.Config.OIDCClientAssertion != nil {
+		authMethod.Config.OIDCClientSecret = ""
+	}
```

Comment:
I chased this method up to `Get` and up to the RPC handler and it looks like this mutates an object we got out of the state store without a copy. Mutating an object in a call named `Get` that isn't even returning that object seems like spooky action at a distance as well. (We do this in a couple places in the state store but only to return a copy of the object "denormalized", and it's heavily documented).

Reply:
Yikes, nice catch! This would be more in keeping with the intent:

```go
clientSecret := authMethod.Config.OIDCClientSecret
if authMethod.Config.OIDCClientAssertion != nil {
	clientSecret = ""
}
```

then refer to the new `clientSecret` var down below in `oidc.NewConfig()`

Think that covers it?

Reply:
Yeah, that works!